libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/barrier.h

```diff
@@ -641,176 +678,423 @@ _LIBCUDACXX_INLINE_VISIBILITY constexpr bool __unused(_Ty...) {return true;}
-                    case 3: __is_async = __memcpy_async_impl<4>::__copy(__destination, __source, __size, __group.thread_rank(), __group.size()); break;
-                    case 2: // fallthrough
-                    case 1: __is_async = __memcpy_async_impl<1>::__copy(__destination, __source, __size, __group.thread_rank(), __group.size()); break;
-                }
-            }
-            else {
-                __is_async = __memcpy_async_impl<_Native_alignment>::__copy(__destination, __source, __size, __group.thread_rank(), __group.size());
+// The __get_size_align struct provides a way to interrogate the guaranteed
+// "alignment" of a provided size. In this case, an n-byte aligned size means
+// that the size is a multiple of n.
+//
+// Use as follows:
+// static_assert(__get_size_align<size_t>::align == 1)
+// static_assert(__get_size_align<aligned_size_t<n>>::align == n)
+template< class... >
+using __custom_void_t = void;
+
+// Default impl: always returns 1.
+template <typename, typename = void>
+struct __get_size_align {
+    static constexpr int align = 1;
+};
+
+// aligned_size_t<n> overload: return n.
+template <typename T>
+struct __get_size_align<T, __custom_void_t<decltype(T::align)>> {
+    static constexpr int align = T::align;
+};
+
+/***********************************************************************
+ * cuda::memcpy_async dispatch
+ *
+ * The dispatch mechanism takes all the arguments and dispatches to the
+ * fastest asynchronous copy mechanism available.
+ *
+ * It returns a __completion_mechanism that indicates which completion mechanism
+ * was used by the copy mechanism. This value can be used by the sync object to
+ * further synchronize if necessary.
+ *
+ ***********************************************************************/
+
+template<_CUDA_VSTD::size_t _Align, typename _Group>
+inline __device__
+__completion_mechanism __dispatch_memcpy_async_any_to_any(_Group const & __group, char * __dest_char, char const * __src_char, _CUDA_VSTD::size_t __size, uint32_t __allowed_completions, uint64_t* __bar_handle) {
+    __cp_async_fallback<_Align>(__group, __dest_char, __src_char, __size);
+    return __completion_mechanism::__sync;
+}
+
+template<_CUDA_VSTD::size_t _Align, typename _Group>
+inline __device__
+__completion_mechanism __dispatch_memcpy_async_global_to_shared(_Group const & __group, char * __dest_char, char const * __src_char, _CUDA_VSTD::size_t __size, uint32_t __allowed_completions, uint64_t* __bar_handle) {
+    NV_IF_TARGET(NV_PROVIDES_SM_90, (
+        const bool __can_use_complete_tx = __allowed_completions & uint32_t(__completion_mechanism::__mbarrier_complete_tx);
+        _LIBCUDACXX_DEBUG_ASSERT(__can_use_complete_tx == (nullptr != __bar_handle), "Pass non-null bar_handle if and only if can_use_complete_tx.");
+        if _LIBCUDACXX_CONSTEXPR_AFTER_CXX11 (16 <= _Align) {
```

Comment:
This can only be constexpr with c++17 onwards
```suggestion
        if _LIBCUDACXX_CONSTEXPR_AFTER_CXX14 (_Align >= 16) {
```

Reply:
Can you explain what you mean by your comment? I currently do not understand the difference between the two macros.